File: lib/core/routing/geo.dart
import 'dart:math';

/// Haversine distance in meters between two geographic points.
/// Direct port from V4 Python engine geo.py.
double haversineM(double lat1, double lng1, double lat2, double lng2) {
  const r = 6371000.0;
  final p1 = _radians(lat1);
  final p2 = _radians(lat2);
  final dp = _radians(lat2 - lat1);
  final dl = _radians(lng2 - lng1);
  final a =
      sin(dp / 2) * sin(dp / 2) + cos(p1) * cos(p2) * sin(dl / 2) * sin(dl / 2);
  return 2 * r * atan2(sqrt(a), sqrt(1 - a));
}

double _radians(double degrees) => degrees * pi / 180.0;

File: engine_files/routing_engine/routing_engine_v4/routing_engine/geo.py
from math import radians, sin, cos, sqrt, atan2

def haversine_m(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    r = 6371000.0
    p1, p2 = radians(lat1), radians(lat2)
    dp = radians(lat2-lat1)
    dl = radians(lng2-lng1)
    a = sin(dp/2)**2 + cos(p1)*cos(p2)*sin(dl/2)**2
    return 2*r*atan2(sqrt(a), sqrt(1-a))
